### ISSUE TITLE
Curate SFT-9 dataset mixes

### ISSUE THREAD

**Issue by olliestanley** [LAION-AI/Open-Assistant#3144]
Iterate on the SFT-8 dataset mixes to create pretraining and final SFT mixes for SFT-9. This requires investigating the quality and usefulness of the datasets. Community input welcome below. See the `sft8_training` [branch](https://github.com/LAION-AI/Open-Assistant/tree/sft8_training) for the code state corresponding to the below SFT-8 configs.

SFT-8 pretraining mix: https://huggingface.co/OpenAssistant/llama-30b-pre-v8-13k-steps/blob/main/README.md

SFT-8 final SFT mix: https://huggingface.co/OpenAssistant/llama-30b-sft-v8-2.5k-steps/blob/main/README.md

Leading on this: @0x22almostEvil

**Comment by 0x22almostEvil** [comment]
I'm here!

**Comment by r7l** [comment]
Would it be possible to use the [Starcoder datasets](https://huggingface.co/datasets/bigcode/starcoderdata)?

**Comment by olliestanley** [comment]
> Would it be possible to use the [Starcoder dataset](https://huggingface.co/datasets/bigcode/starcoderdata)?

Currently we include some RedPajama data with a language modelling objective during SFT to try to prevent catastrophic forgetting of pretraining knowledge. Maybe it would be possible to do something similar with StarCoder data. But I don't think we could train on the whole dataset, that would just be hugely expensive and more in the realms of foundation model pretraining than assistant finetuning.

**Comment by r7l** [comment]
Understandable. I'd assume that a large portion of current OA users are coders. So it might be reasonable for the start to have a good understanding for coding. It's pretty ok for SFT-7 already but there will always be room for improvement.

**Comment by 0x22almostEvil** [comment]
Yeah, we might collect some good datasets for coding as well

**Comment by olliestanley** [comment]
Since the results of the Guanaco paper I think it is clear SFT-9 should use a much smaller set of finetuning data and focus on high quality. I suggest we try a run dropping synthetic datasets, with maybe exceptions for those synthetic datasets which are clearly high-quality

**Comment by andreaskoepf** [comment]
While we definitely should use QLoRA (a groundbreaking result for the whole ML community) and only try a super-high quality final fine-tuning run (like OA top-1 threads, i.e. as it was done for Guanaco) I think the total situation is not that simple.

In the fine-print of his twitter [mega-thread](https://twitter.com/Tim_Dettmers/status/1661379354507476994) Tim Dettmers writes:

- "Our main finding here: (1) instruction tuning datasets are good for instruction following but bad for chatbot performance; (2) you can create a 99.3% of ChatGPT performance level chatbot with QLoRA in just 24 hours of fine-tuning!" ([tweet](https://twitter.com/Tim_Dettmers/status/1661379410719539211) -> i.e instruction following and "fluffy" chat are two different things
- "its really bad at math" ([tweet](https://twitter.com/Tim_Dettmers/status/1661380027408166914)

What we clearly see is that the style of the model output can be greatly modified already with 1k (Lima) or 10k (QLoRA) examples. Whether additional "pre-training" is beneficial for capabilities or not was IMO not analyzed. We observed that pre-training has clearly an influence (e.g. negative with prosocial and positive with grade-school-math).
Also we know that our SFT7e3 model although it most of the time fails to generate rhyming poems is our best model for following instructions and handling plugin-requests.

I see two obvious solutions:
- use something "mode" in system-prompt to specify whether we want instruction following or fluffy-talk mode
- use multiple specialized models, e.g. one for chat and another one for instruction following

**Comment by olliestanley** [comment]
I agree that there is a clear distinction between datasets useful for chat-tuning vs instruction-following, but have a few points here.

> We already followed a 2-stage training approach. Guanaco of course goes a step further and trained only on the highest quality OA-data. When we decided to use the full OA set (I implemented a top-k thread filter which was not used) the idea was to create a diverse SFT output as input to the RL stage.

This makes sense, but it seems to me that even if we continue the 2-stage approach we can most likely get sufficiently diverse outputs for RL even with a highly filtered OA set.

> In the fine-print of his twitter [mega-thread](https://twitter.com/Tim_Dettmers/status/1661379354507476994) Tim Dettmers writes:
> 
> * "Our main finding here: (1) instruction tuning datasets are good for instruction following but bad for chatbot performance; (2) you can create a 99.3% of ChatGPT performance level chatbot with QLoRA in just 24 hours of fine-tuning!" ([tweet](https://twitter.com/Tim_Dettmers/status/1661379410719539211)) -> i.e instruction following and "fluffy" chat are two different things
> * "its really bad at math" ([tweet](https://twitter.com/Tim_Dettmers/status/1661380027408166914))

Yes, my suggestion would be to retain these high-quality instruction-following datasets (e.g. math instructions, poetry instructions, coding instructions, and I believe Dragan is building a plugin-instruction dataset) but remove the synthetic chat datasets. It seems like we do not need Alpaca, Vicuna, WizardLM, prosocial, roleplay, etc datasets which are chat-focused and likely to be lower quality than filtered OA and Dolly data, perhaps with the exception of alpaca_gpt4?

There are some datasets I haven't looked at so am less sure about (OIG file, soda, webgpt, recipes, wiki QA datasets)

> I see two obvious solutions/approaches for chat vs. plugins:
> 
> * use something "mode" in system-prompt to specify whether we want instruction following or fluffy-talk mode
> * use multiple specialized models, e.g. one for chat and another one for instruction following

I personally prefer the idea of having a single model which can do both - it aligns much better with the OA vision of running on consumer hardware. So the system prompt idea seems like a good starting point, imo

**Comment by draganjovanovich** [comment]
I am for single mode/model approach. And completely removing instruction datasets seams a bit too much. But I am all in for leaving only top quality samples/datasets.
